pkg/scheduler/core/equivalence/eqivalence.go

```diff
@@ -63,15 +63,57 @@ func NewCache() *Cache {
 type NodeCache struct {
 	mu    sync.RWMutex
 	cache predicateMap
+	// generations stores generation numbers for predicates, incremented on
+	// predicate invalidation. Created on first update. Use 0 if does not
+	// exist.
+	generations map[string]uint64
 }
 
 // newNodeCache returns an empty NodeCache.
 func newNodeCache() *NodeCache {
 	return &NodeCache{
-		cache: make(predicateMap),
+		cache:       make(predicateMap),
+		generations: make(map[string]uint64),
 	}
 }
 
+// NodeCacheSnapshot represents a snapshot of NodeCache reference and generations.
+type NodeCacheSnapshot struct {
+	NodeCache   *NodeCache
```

Comment:
Why do we need to store this pointer? We just need to snapshot the generation number.

Reply:
I have some explanations [here](https://github.com/kubernetes/kubernetes/pull/67308#issuecomment-414054511). For invalidating node, deleting node cache is not enough, because node is invalidated before `lookupResult`, a fresh node cache will be created in `lookupResult` and stale data will be written into it. To fix this, we can get node cache before snapshotting scheduler cache. If in-flight predicate checker write stale data into it, node cache itself can be removed from equivalence cache on invalidation, so scheduler will use fresh node cache next run.

I prefer this because 1) we don't need to store node cache forever even after node is deleted from cluster 2) checkNode needs to get node cache one by one, getting node cache pointers all at once does not add performance overhead.

Reply:
@misterikkit @resouer What do you think?